invalid c generated for my complex `arrays.fold` call

### Describe the bug

Code snippet that seems to cause problems is in `zulip/bot.v` lines 66:88

Didn't look into exactly what is causing the cgen issue, maybe a smaller repro can be made.
```v
		arrays.fold(r.events, []Event{}, fn (acc []Event, a json2.Any) []Event {
			typ := json2.decode[ResponseStub](a.str()) or { ResponseStub {.unknown} }.etyp
			res := match typ {
				.message {
					Event(json2.decode[MessageEvent](a.str()) or {
						panic(err)
					})
				}
				.reaction {
					Event(json2.decode[ReactionEvent](a.str()) or {
						panic(err)
					})
				}
				else {
					eprintln('UNSUPPORTED EVENT TYPE')
					none
				}
			}
			if r := res {
				return arrays.append(acc, [r])
			}
			return acc
		})
```

```
❯ v run echobot.v
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/echobot.01KGDYAXWWYWK4TXYP1JTFQD4A.tmp.c:6504: warning: type defaults to int
cc: /tmp/v_1000/echobot.01KGDYAXWWYWK4TXYP1JTFQD4A.tmp.c:6504: error: ';' expected (got "*")
=================================================================
Try passing `-g` when compiling, to see a .v file:line information, that correlates more with the C error.
(Alternatively, pass `-show-c-output`, to print the full C error message).
builder error:
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

### Reproduction Steps

See: https://github.com/dy-tea/vulip

```sh
git clone https://github.com/dy-tea/vulip
cd vulip
v run echobot.v
```

### Expected Behavior

Should compile, or maybe not; I am not entirely sure what I was doing here.

### Current Behavior

.

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.5.0 50a7d53

### Environment details (OS name and version, etc.)

|V full version      |V 0.5.0 a965a390eb6db621df3a6133c5949459b169cd36.50a7d53
|:-------------------|:-------------------
|OS                  |linux, "CachyOS"
|Processor           |12 cpus, 64bit, little endian, 13th Gen Intel(R) Core(TM) i5-1335U
|Memory              |4.19GB/15.31GB
|                    |
|V executable        |/home/dylan/Repos/v/v
|V last modified time|2026-01-28 17:56:57
|                    |
|V home dir          |OK, value: /home/dylan/Repos/v
|VMODULES            |OK, value: /home/dylan/.vmodules
|VTMP                |OK, value: /tmp/v_1000
|Current working dir |OK, value: /home/dylan
|                    |
|Git version         |git version 2.52.0
|V git status        |weekly.2026.02-182-g50a7d536
|.git/config present |true
|                    |
|cc version          |cc (GCC) 15.2.1 20260103
|gcc version         |gcc (GCC) 15.2.1 20260103
|clang version       |clang version 21.1.6
|tcc version         |tcc version 0.9.28rc 2025-02-13 HEAD@f8bd136d (x86_64 Linux)
|tcc git status      |thirdparty-linux-amd64 696c1d84
|emcc version        |N/A
|glibc version       |ldd (GNU libc) 2.42

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -754,9 +754,15 @@ fn (mut g Gen) gen_anon_fn_decl(mut node ast.AnonFn) {
 	}
 	pos := g.out.len
 	was_anon_fn := g.anon_fn
+	prev_stmt_path_pos := g.stmt_path_pos.clone()
+	prev_skip_stmt_pos := g.skip_stmt_pos
+	g.stmt_path_pos = []
+	g.skip_stmt_pos = false
 	g.anon_fn = node
 	g.fn_decl(node.decl)
 	g.anon_fn = was_anon_fn
+	g.skip_stmt_pos = prev_skip_stmt_pos
+	g.stmt_path_pos = prev_stmt_path_pos
 	builder.write_string(g.out.cut_to(pos))
 	out := builder.str()
 	g.anon_fn_definitions << out

--- a/vlib/v/tests/fns/anon_fn_in_generic_call_with_match_and_inner_generic_call_test.v
+++ b/vlib/v/tests/fns/anon_fn_in_generic_call_with_match_and_inner_generic_call_test.v
@@ -1,0 +1,41 @@
+module main
+
+import arrays
+import x.json2
+
+enum EventType {
+	message
+	reaction
+	unknown
+}
+
+struct ResponseStub {
+	etyp EventType @[json: 'type']
+}
+
+fn count_messages(events []json2.Any) !int {
+	return if events.len > 0 {
+		arrays.fold(events, 0, fn (acc int, a json2.Any) int {
+			typ := json2.decode[ResponseStub](a.str()) or { ResponseStub{.unknown} }.etyp
+			res := match typ {
+				.message { 1 }
+				.reaction { 2 }
+				else { none }
+			}
+			if r := res {
+				return acc + r
+			}
+			return acc
+		})
+	} else {
+		error('empty')
+	}
+}
+
+fn test_fix_issue_26498() {
+	events := [
+		json2.Any('{"type":"message"}'),
+		json2.Any('{"type":"reaction"}'),
+	]
+	assert count_messages(events)! == 3
+}

